src/ctap/credential_id.rs

```diff
@@ -0,0 +1,475 @@
+
+pub const CBOR_CREDENTIAL_ID_VERSION: u8 = 0x01;
+
+pub const MAX_PADDING_LENGTH: u8 = 0xBF;
+
+// Data fields that are contained in the credential ID of non-discoverable credentials.
+struct CredentialSource {
+    private_key: PrivateKey,
+    rp_id_hash: [u8; 32],
+    cred_protect_policy: Option<CredentialProtectionPolicy>,
+}
+
+// The data fields contained in the credential ID are serizlied using CBOR maps.
+// Each field is associated with a unique tag, implemented with a CBOR unsigned key.
+enum CredentialSourceField {
+    PrivateKey = 0,
+    RpIdHash = 1,
+    CredProtectPolicy = 2,
+}
+
+impl From<CredentialSourceField> for sk_cbor::Value {
+    fn from(field: CredentialSourceField) -> sk_cbor::Value {
+        (field as u64).into()
+    }
+}
+
+fn decrypt_legacy_credential_id(
+    env: &mut impl Env,
+    bytes: &[u8],
+) -> Result<Option<CredentialSource>, Ctap2StatusCode> {
+    let aes_enc_key = crypto::aes256::EncryptionKey::new(&env.key_store().key_handle_encryption()?);
+    let plaintext = aes256_cbc_decrypt(&aes_enc_key, bytes, true)?;
+    if plaintext.len() != 64 {
+        return Ok(None);
+    }
+    let private_key = if let Some(key) = PrivateKey::new_ecdsa_from_bytes(&plaintext[..32]) {
+        key
+    } else {
+        return Ok(None);
+    };
+    Ok(Some(CredentialSource {
+        private_key,
+        rp_id_hash: plaintext[32..64].try_into().unwrap(),
+        cred_protect_policy: None,
+    }))
+}
+
+fn decrypt_cbor_credential_id(
+    env: &mut impl Env,
+    bytes: &[u8],
+) -> Result<Option<CredentialSource>, Ctap2StatusCode> {
+    let aes_enc_key = crypto::aes256::EncryptionKey::new(&env.key_store().key_handle_encryption()?);
+    let mut plaintext = aes256_cbc_decrypt(&aes_enc_key, bytes, true)?;
+    remove_padding(&mut plaintext)?;
+
+    let cbor_credential_source = cbor_read(plaintext.as_slice())?;
+    destructure_cbor_map! {
+      let {
+          CredentialSourceField::PrivateKey => private_key,
+          CredentialSourceField::RpIdHash=> rp_id_hash,
+          CredentialSourceField::CredProtectPolicy => cred_protect_policy,
+      } = extract_map(cbor_credential_source)?;
+    }
+    Ok(match (private_key, rp_id_hash, cred_protect_policy) {
+        (Some(private_key), Some(rp_id_hash), cred_protect_policy) => {
+            let private_key = PrivateKey::try_from(private_key)?;
+            let rp_id_hash = extract_byte_string(rp_id_hash)?;
+            if rp_id_hash.len() != 32 {
+                return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+            }
+            let cred_protect_policy = if let Some(policy) = cred_protect_policy {
+                Some(CredentialProtectionPolicy::try_from(policy)?)
+            } else {
+                None
+            };
+            Some(CredentialSource {
+                private_key,
+                rp_id_hash: rp_id_hash.try_into().unwrap(),
+                cred_protect_policy,
+            })
+        }
+        _ => None,
```

Comment:
I double checked the `Ok(None)` vs `Err(_)` question, and this looked interesting, everything is is correct I believe. If the private key or hash are missing, that would be a bad sign indeed, maybe this needs to be an error?